sdks/java/extensions/timeseries/src/main/java/org/apache/beam/sdk/extensions/timeseries/joins/BiTemporalStreams.java

```diff
@@ -0,0 +1,462 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.timeseries.joins;
+
+import java.util.*;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.state.*;
+import org.apache.beam.sdk.state.Timer;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.*;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This sample will take two streams left stream and right stream. It will match the left stream to
+ * the nearest right stream based on their timestamp. Nearest left stream is either <= to the right
+ * stream timestamp.
+ *
+ * <p>If two values in the right steam have the same timestamp then the results are
+ * non-deterministic.
+ */
+@Experimental
+public class BiTemporalStreams {
+
+  public static <K, V1, V2> BiTemporalJoin<K, V1, V2> join(
+      TupleTag<V1> leftTag, TupleTag<V2> rightTag, Duration window) {
+    return new BiTemporalJoin<>(leftTag, rightTag, window);
+  }
+
+  @Experimental
+  public static class BiTemporalJoin<K, V1, V2>
+      extends PTransform<KeyedPCollectionTuple<K>, PCollection<BiTemporalJoinResult<K, V1, V2>>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BiTemporalStreams.class);
+
+    // Sets the limit at which point the processed left stream values are garbage collected
+    static int GC_LIMIT = 1000;
```

Comment:
OOO at the moment, but the current experimental implementation does not deal with late data ( I think I had a note in the java doc). When I am back I will add a check to throw illegal state if late data is not set to ZERO.  Until we have retractions the best way that we can deal with late data is to deadletter it.

The GC_LIMIT ( which should have a better name, will fix that when I am back) is actually to reduce the amount of processing needed in sorting the left hand side. As the whole list needs to be sorted across each finish bundle, but once an element is processed it is no longer needed. Once we have Sorted Map State most of the code for this class can be deleted, as there will be no need for the cache that we have to use at the moment.  :-)

Reply:
With regards to Je-ik first comment, 
@amaliujia @kennknowles had a quick glance at the StatefullDoFnRunner, it uses a Timer to do GC. I dont think all runners use Timers for this operation, but as je-ik mentioned some will do ... If I recall there is some issues with how multiple timers will interact with each other, when a Timer resets itself. Should this be standardized, as I think right now this could potentially mean things will work on some runners and not on others?

Reply:
@rezarokni I have managed to (hopefully finally) fix the issue with timers. Related PR is #9190, I cherry picked the two commits into PR rezarokni#2 and the failing tests are green on my local machine, with the exception of `BiTemporalCacheTest.cacheTest`, which is still failing (consistently). Don't know what is the issue with that, but I hope it is no longer related to the timer ordering. Could you validate that?

Reply:
Thanx for looking into this Jan, just back from OOO will look at this over the next week.

Reply:
Looks like one of the keys is no longer getting processed. Thats the symptom, digging into the cause.